Add option to use provided conf.py

kmorel opened the pull request:
Some projects will want to provide their own `conf.py` rather than have one built. This change adds a `CONF_FILE` option to provide such a file.

Because configuration such as breathe needs to be added to the file, it is copied to the binary directory, and that version is used in the build.

An additional feature is that CMake variable substitution is performed on the configure file. This makes it easy to add system-specific configuration such as absolute paths to the configuration.

kmorel commented:
@k0ekk0ek, I noticed that your FindSphinx.cmake had a todo item to allow code to provide their own `conf.py` file. That was something I could use, so I added a simple feature for it. Does this implementation make sense?

k0ekk0ek commented:
Hi @kmorel. Always nice to hear someone finds this little module helpful :slightly_smiling_face: I haven't looked at the code in a while, but I did implement this for another project at some point. I dug up the code from there.

```diff
diff --git a/cmake/Modules/FindSphinx.cmake b/cmake/Modules/FindSphinx.cmake
index 0b3e8c0..0f5c1fc 100644
--- a/cmake/Modules/FindSphinx.cmake
+++ b/cmake/Modules/FindSphinx.cmake
@@ -354,12 +354,14 @@ function(sphinx_add_docs _target)
 
   set(_cachedir "${CMAKE_CURRENT_BINARY_DIR}/${_target}.cache")
   file(MAKE_DIRECTORY "${_cachedir}")
-  file(MAKE_DIRECTORY "${_cachedir}/_static")
+  file(COPY "${_sourcedir}/_static" DESTINATION "${_cachedir}")
+  file(COPY "${_sourcedir}/_templates" DESTINATION "${_cachedir}")
 
-  # FIXME: This step must become optional! Based on existance of conf.py in
-  #        the source directory and whether or not sphinx-quickstart was
-  #        found. If both are missing, it's an error!
-  _Sphinx_generate_confpy(${_target} "${_cachedir}")
+  if(EXISTS "${_sourcedir}/conf.py")
+    configure_file("${_sourcedir}/conf.py" "${_cachedir}/conf.py" @ONLY)
+  else()
+    _Sphinx_generate_confpy(${_target} "${_cachedir}")
+  endif()
 
   if(_breathe_projects)
     file(APPEND "${_cachedir}/conf.py"
```

I like that it's more-or-less automatic in the patch, but I guess we can add an option to make it explicit too. In which case copying of the directory should be optional too I guess(?) I'll have a look after the weekend.

k0ekk0ek committed:
Use existing conf.py if provided

Rather than always creating a conf.py, check to see if a conf.py
exists and use that if possible.

Also use the project's _source and _tempalte directories if they
exist.

kmorel commented:
Yes, I really appreciate being able to use your module. It has saved me quite a bit of time setting up a Sphinx project.

Your method of finding `conf.py` works for me. I've changed this PR to use your patch.

I did make a change to the way the patch copied the `_static` and `_template` directories. The patch would fail if either directory did not exist, so I added a check before doing the copy.

Does this PR look reasonable to merge?

k0ekk0ek commented on cmake/Modules/FindSphinx.cmake:
```diff
@@ -354,12 +354,18 @@ function(sphinx_add_docs _target)
-  file(MAKE_DIRECTORY "${_cachedir}/_static")
+  if(EXISTS "${_sourcedir}/_static")
+    file(COPY "${_sourcedir}/_static" DESTINATION "${_cachedir}")
+  endif()
```
Should we add an `else()` that does `MAKE_DIRECTORY`?

kmorel replied:
Sure. Pushed.

k0ekk0ek commented:
Changes look good. Thanks @kmorel!
